File: neurolab-lite/frontend/src/engines/aiEngine.js
/**
 * Feedforward Neural Network forward pass (client-side, pure JS).
 *
 * Replaces PyTorch backend with simple matrix operations.
 * Uses a seeded PRNG for reproducible Xavier-uniform weight initialization.
 */

// Simple seeded PRNG (mulberry32)
function mulberry32(seed) {
  return function () {
    seed |= 0;
    seed = (seed + 0x6d2b79f5) | 0;
    let t = Math.imul(seed ^ (seed >>> 15), 1 | seed);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

const ACTIVATIONS = {
  relu: (x) => Math.max(0, x),
  sigmoid: (x) => 1 / (1 + Math.exp(-x)),
  tanh: (x) => Math.tanh(x),
};

/**
 * Xavier-uniform initialization: U(-limit, limit) where limit = sqrt(6 / (fan_in + fan_out))
 */
function xavierMatrix(rows, cols, rng) {
  const limit = Math.sqrt(6 / (rows + cols));
  const mat = [];
  for (let r = 0; r < rows; r++) {
    const row = [];
    for (let c = 0; c < cols; c++) {
      row.push(rng() * 2 * limit - limit);
    }
    mat.push(row);
  }
  return mat;
}

/**
 * Matrix-vector multiply: result[i] = sum_j(W[i][j] * x[j]) + bias[i]
 */
function linearForward(W, bias, input) {
  const outSize = W.length;
  const result = [];
  for (let i = 0; i < outSize; i++) {
    let sum = bias[i];
    for (let j = 0; j < input.length; j++) {
      sum += W[i][j] * input[j];
    }
    result.push(sum);
  }
  return result;
}

/**
 * Batched forward pass: process each sample independently.
 */
export function runForwardPass({ config, input_data }) {
  const {
    input_size = 2,
    hidden_sizes = [4, 4],
    output_size = 1,
    activation = 'relu',
  } = config;

  const actFn = ACTIVATIONS[activation] || ACTIVATIONS.relu;
  const rng = mulberry32(42);

  // Build layers: list of { W, bias, isOutput }
  const layers = [];
  let prevSize = input_size;
  for (const hs of hidden_sizes) {
    layers.push({
      W: xavierMatrix(hs, prevSize, rng),
      bias: new Array(hs).fill(0),
      isOutput: false,
    });
    prevSize = hs;
  }
  layers.push({
    W: xavierMatrix(output_size, prevSize, rng),
    bias: new Array(output_size).fill(0),
    isOutput: true,
  });

  // Process each sample in the batch
  const batchOutput = [];
  const layerActivations = [{ layer_name: 'Input', values: input_data }];
  // We'll accumulate per-layer values across the batch
  const layerBatchValues = layers.map(() => []);

  const architecture = [`Input: ${input_size} features`];

  for (const sample of input_data) {
    let x = sample;
    for (let li = 0; li < layers.length; li++) {
      const { W, bias, isOutput } = layers[li];
      x = linearForward(W, bias, x);
      if (!isOutput) {
        x = x.map(actFn);
      }
      layerBatchValues[li].push(x);
    }
    batchOutput.push(x);
  }

  // Build layer activations and architecture descriptions
  let hiddenIdx = 0;
  for (let li = 0; li < layers.length; li++) {
    const { W, isOutput } = layers[li];
    const inFeatures = W[0].length;
    const outFeatures = W.length;
    if (isOutput) {
      layerActivations.push({
        layer_name: 'Output',
        values: layerBatchValues[li],
      });
      architecture.push(`Output: Linear(${inFeatures} → ${outFeatures})`);
    } else {
      hiddenIdx++;
      layerActivations.push({
        layer_name: `Hidden ${hiddenIdx}`,
        values: layerBatchValues[li],
      });
      architecture.push(
        `Hidden ${hiddenIdx}: Linear(${inFeatures} → ${outFeatures}) + ${activation}`
      );
    }
  }

  return {
    output: batchOutput,
    layer_activations: layerActivations,
    architecture,
  };
}

File: neurolab-lite/backend/engines/ai_engine.py
"""
AI (PyTorch) Feedforward Neural Network Engine.

This module implements a simple feedforward artificial neural network using
PyTorch, designed for educational comparison with biological neuron models.

BIOLOGICAL vs ARTIFICIAL NEURON — Key Differences:
┌───────────────────┬──────────────────────┬──────────────────────┐
│ Property          │ Biological (LIF)     │ Artificial (ANN)     │
├───────────────────┼──────────────────────┼──────────────────────┤
│ Communication     │ Spikes (binary)      │ Continuous values    │
│ Temporal dynamics │ Yes (differential    │ No (instantaneous)   │
│                   │ equations over time) │                      │
│ Learning rule     │ Hebbian / STDP       │ Backpropagation      │
│ Activation        │ Threshold → spike    │ ReLU, sigmoid, etc.  │
│ Energy            │ ~20W (whole brain)   │ ~300W (GPU)          │
│ Computation       │ Asynchronous         │ Synchronous layers   │
└───────────────────┴──────────────────────┴──────────────────────┘

The forward pass computes:
    h_1 = activation(W_1 @ x + b_1)
    h_2 = activation(W_2 @ h_1 + b_2)
    ...
    output = W_n @ h_{n-1} + b_n
"""

import torch
import torch.nn as nn

from models.ai import AIModelConfig


def build_network(config: AIModelConfig) -> nn.Sequential:
    """
    Build a PyTorch Sequential feedforward network from config.

    Each hidden layer consists of:
    1. Linear transformation: y = Wx + b
    2. Activation function (models the neuron's "firing rate" response)

    The output layer has no activation — it produces raw values.
    """
    activation_map = {
        "relu": nn.ReLU,       # max(0, x) — most common in modern deep learning
        "sigmoid": nn.Sigmoid, # 1/(1+e^{-x}) — historically inspired by biological firing rates
        "tanh": nn.Tanh,       # Centered version of sigmoid, range [-1, 1]
    }

    act_fn = activation_map.get(config.activation, nn.ReLU)

    layers = []
    prev_size = config.input_size

    for hidden_size in config.hidden_sizes:
        layers.append(nn.Linear(prev_size, hidden_size))
        layers.append(act_fn())
        prev_size = hidden_size

    # Output layer — no activation (raw logits)
    layers.append(nn.Linear(prev_size, config.output_size))

    return nn.Sequential(*layers)


def run_forward_pass(
    config: AIModelConfig,
    input_data: list[list[float]],
) -> dict:
    """
    Run a forward pass and capture activations at every layer.

    This is analogous to recording voltage traces from biological neurons,
    except here we capture the continuous-valued activations at each layer.

    Args:
        config: Network architecture specification.
        input_data: Input samples as list of lists.

    Returns:
        Dictionary with 'output', 'layer_activations', 'architecture'.
    """
    model = build_network(config)

    # Use eval mode and no gradients — we're only doing inference
    model.eval()

    # Initialize weights to small random values for reproducibility
    torch.manual_seed(42)
    for param in model.parameters():
        if param.dim() > 1:
            nn.init.xavier_uniform_(param)

    input_tensor = torch.tensor(input_data, dtype=torch.float32)

    layer_activations = []
    architecture = []

    # Record input
    layer_activations.append({
        "layer_name": "Input",
        "values": input_tensor.tolist(),
    })
    architecture.append(f"Input: {config.input_size} features")

    # Forward pass with intermediate recording
    x = input_tensor
    layer_idx = 0
    with torch.no_grad():
        for i, layer in enumerate(model):
            x = layer(x)
            if isinstance(layer, nn.Linear):
                layer_idx += 1
                is_output = (i == len(model) - 1)
                name = "Output" if is_output else f"Hidden {layer_idx}"
                layer_activations.append({
                    "layer_name": name,
                    "values": x.tolist(),
                })
                if is_output:
                    architecture.append(
                        f"Output: Linear({layer.in_features} → {layer.out_features})"
                    )
                else:
                    architecture.append(
                        f"Hidden {layer_idx}: Linear({layer.in_features} → {layer.out_features}) + {config.activation}"
                    )

    return {
        "output": x.tolist(),
        "layer_activations": layer_activations,
        "architecture": architecture,
    }
